File: apps/ios/BrickBackKit/Sources/BrickBackKit/Local/AppDatabase.swift
import Foundation
import GRDB

/// On-device local-first store — the **source of truth** for the user. Cloud (BrickBack
/// user project) is only a premium sync mirror. Direct port of `app_database.dart`
/// (Drift schemaVersion 3), minus the `step_qty` column (00-architecture §5): the per-part
/// counting step is in-memory session state (S3), never persisted.
///
/// All writes funnel through `writer` (a serial GRDB queue) so the UI and the later sync
/// engine never fight over the connection (S1 risk note).
public final class AppDatabase: Sendable {
    /// Serial write queue. Package-internal: the app never touches GRDB directly — it goes
    /// through the repositories, which expose GRDB-free async APIs.
    let writer: any DatabaseWriter

    init(_ writer: any DatabaseWriter) throws {
        self.writer = writer
        try Self.migrator.migrate(writer)
    }

    /// The on-disk store at `Application Support/brickback.sqlite`.
    public static func live() throws -> AppDatabase {
        let fm = FileManager.default
        let dir = try fm.url(for: .applicationSupportDirectory, in: .userDomainMask, appropriateFor: nil, create: true)
        let url = dir.appendingPathComponent("brickback.sqlite")
        var config = Configuration()
        config.foreignKeysEnabled = false // parity with Drift (no FKs declared; parents pushed first)
        let queue = try DatabaseQueue(path: url.path, configuration: config)
        return try AppDatabase(queue)
    }

    /// In-memory store for tests (mirrors `AppDatabase.forTesting`).
    public static func inMemory() throws -> AppDatabase {
        try AppDatabase(try DatabaseQueue())
    }

    // MARK: - Migrations

    /// Named migrations reproducing the Drift history (00-architecture §5):
    /// - v1: rebuild_sets, rebuild_parts (no category_name), rebuild_minifigs, verifications
    /// - v2: + rebuild_parts.category_name; create rebuild_extra_parts
    /// - v3: + rebuild_sets.images_cached_at (S9 offline image cache; device-local, not synced).
    ///   This is **not** the Drift v3 `step_qty` column — that stays dropped (00-architecture §5).
    static var migrator: DatabaseMigrator {
        var migrator = DatabaseMigrator()

        migrator.registerMigration("v1") { db in
            try db.create(table: "rebuild_sets") { t in
                t.column("id", .text).primaryKey()
                t.column("set_item_id", .integer).notNull()
                t.column("name", .text).notNull().defaults(to: "")
                t.column("theme", .text)
                t.column("year", .integer)
                t.column("image_url", .text)
                t.column("total_parts", .integer).notNull().defaults(to: 0)
                t.column("verified_at", .datetime)
                t.column("created_at", .datetime).notNull().defaults(sql: "CURRENT_TIMESTAMP")
                t.column("updated_at", .datetime).notNull().defaults(sql: "CURRENT_TIMESTAMP")
                t.column("dirty", .boolean).notNull().defaults(to: true)
                t.column("deleted", .boolean).notNull().defaults(to: false)
            }

            try db.create(table: "rebuild_parts") { t in
                t.column("rebuild_set_id", .text).notNull()
                t.column("part_item_id", .integer).notNull()
                t.column("color_id", .integer).notNull()
                t.column("needed_qty", .integer).notNull().defaults(to: 0)
                t.column("have_qty", .integer).notNull().defaults(to: 0)
                t.column("part_name", .text).notNull().defaults(to: "")
                t.column("part_num", .text)
                t.column("part_cat_id", .integer)
                // category_name added in v2
                t.column("color_name", .text)
                t.column("color_rgb", .text)
                t.column("image_url", .text)
                t.column("bl_part_id", .text)
                t.column("bl_color_id", .integer)
                t.column("updated_at", .datetime).notNull().defaults(sql: "CURRENT_TIMESTAMP")
                t.column("dirty", .boolean).notNull().defaults(to: true)
                t.column("deleted", .boolean).notNull().defaults(to: false)
                t.primaryKey(["rebuild_set_id", "part_item_id", "color_id"])
            }

            try db.create(table: "rebuild_minifigs") { t in
                t.column("rebuild_set_id", .text).notNull()
                t.column("minifig_item_id", .integer).notNull()
                t.column("needed_qty", .integer).notNull().defaults(to: 0)
                t.column("have_qty", .integer).notNull().defaults(to: 0)
                t.column("name", .text).notNull().defaults(to: "")
                t.column("image_url", .text)
                t.column("updated_at", .datetime).notNull().defaults(sql: "CURRENT_TIMESTAMP")
                t.column("dirty", .boolean).notNull().defaults(to: true)
                t.column("deleted", .boolean).notNull().defaults(to: false)
                t.primaryKey(["rebuild_set_id", "minifig_item_id"])
            }

            try db.create(table: "verifications") { t in
                t.column("id", .text).primaryKey()
                t.column("rebuild_set_id", .text).notNull()
                t.column("set_item_id", .integer).notNull()
                t.column("completion_pct", .double).notNull().defaults(to: 0)
                t.column("parts_needed", .integer)
                t.column("parts_found", .integer)
                t.column("minifigs_needed", .integer)
                t.column("minifigs_found", .integer)
                t.column("flags", .text).notNull().defaults(to: "{}")
                t.column("notes", .text)
                t.column("verified_at", .datetime).notNull().defaults(sql: "CURRENT_TIMESTAMP")
                t.column("updated_at", .datetime).notNull().defaults(sql: "CURRENT_TIMESTAMP")
                t.column("dirty", .boolean).notNull().defaults(to: true)
                t.column("deleted", .boolean).notNull().defaults(to: false)
            }
        }

        migrator.registerMigration("v2") { db in
            // v1 -> v2: "group by type" (category_name) + the extras (spares) list.
            try db.alter(table: "rebuild_parts") { t in
                t.add(column: "category_name", .text)
            }
            try db.create(table: "rebuild_extra_parts") { t in
                t.column("rebuild_set_id", .text).notNull()
                t.column("part_item_id", .integer).notNull()
                t.column("color_id", .integer).notNull()
                t.column("needed_qty", .integer).notNull().defaults(to: 0)
                t.column("have_qty", .integer).notNull().defaults(to: 0)
                t.column("part_name", .text).notNull().defaults(to: "")
                t.column("part_num", .text)
                t.column("part_cat_id", .integer)
                t.column("category_name", .text)
                t.column("color_name", .text)
                t.column("color_rgb", .text)
                t.column("image_url", .text)
                t.primaryKey(["rebuild_set_id", "part_item_id", "color_id"])
            }
        }

        migrator.registerMigration("v3") { db in
            // Offline mode (S9): a device-local marker per set — stamped once every one of the
            // set's image URLs is cached on disk. NULL ⇒ prefetch incomplete ⇒ resume when next
            // online. Never synced (no dirty/deleted); the cloud schema is unchanged.
            try db.alter(table: "rebuild_sets") { t in
                t.add(column: "images_cached_at", .datetime)
            }
        }

        return migrator
    }
}

File: apps/mobile/lib/core/db/app_database.dart
import 'package:drift/drift.dart';
import 'package:drift_flutter/drift_flutter.dart';

part 'app_database.g.dart';

/// On-device local-first store — the **source of truth** for the user.
/// Cloud (BrickBack user project) is only a premium sync mirror.
///
/// Every user row carries sync columns: [id]/composite key == the cloud id,
/// [updatedAt], [dirty] (needs push), [deleted] (tombstone). Metadata snapshot
/// columns let the whole counting UI work fully offline.

@DataClassName('RebuildSetRow')
class RebuildSets extends Table {
  TextColumn get id => text()(); // client-generated uuid (== cloud id)
  IntColumn get setItemId => integer()(); // catalog items.id
  TextColumn get name => text().withDefault(const Constant(''))();
  TextColumn get theme => text().nullable()();
  IntColumn get year => integer().nullable()();
  TextColumn get imageUrl => text().nullable()();
  IntColumn get totalParts => integer().withDefault(const Constant(0))();
  DateTimeColumn get verifiedAt => dateTime().nullable()();
  DateTimeColumn get createdAt => dateTime().withDefault(currentDateAndTime)();
  DateTimeColumn get updatedAt => dateTime().withDefault(currentDateAndTime)();
  BoolColumn get dirty => boolean().withDefault(const Constant(true))();
  BoolColumn get deleted => boolean().withDefault(const Constant(false))();
  // Device-local F4 offline-image completeness marker. NULL ⇒ prefetch
  // incomplete ⇒ resume when next online/opened; stamped once every one of the
  // set's live image URLs is on disk. NEVER synced — the cloud schema is
  // unchanged and push/pull touch explicit column lists that exclude this one
  // (verified in sync_service.dart). v4.
  DateTimeColumn get imagesCachedAt => dateTime().nullable()();

  @override
  Set<Column> get primaryKey => {id};
}

@DataClassName('RebuildPartRow')
class RebuildParts extends Table {
  TextColumn get rebuildSetId => text()();
  IntColumn get partItemId => integer()(); // catalog parts.item_id
  IntColumn get colorId => integer()(); // catalog colors.id
  IntColumn get neededQty => integer().withDefault(const Constant(0))();
  IntColumn get haveQty => integer().withDefault(const Constant(0))();
  // Per-part tap increment for counting. Device-local UX helper (not in the
  // cloud schema), so it's never marked dirty / synced. v3.
  IntColumn get stepQty => integer().withDefault(const Constant(1))();
  // Metadata snapshot (taken from the catalog at add-time; enables offline UI)
  TextColumn get partName => text().withDefault(const Constant(''))();
  TextColumn get partNum => text().nullable()();
  IntColumn get partCatId => integer().nullable()();
  TextColumn get categoryName => text().nullable()(); // v2: enables "group by type"
  TextColumn get colorName => text().nullable()();
  TextColumn get colorRgb => text().nullable()();
  TextColumn get imageUrl => text().nullable()();
  TextColumn get blPartId => text().nullable()();
  IntColumn get blColorId => integer().nullable()();
  DateTimeColumn get updatedAt => dateTime().withDefault(currentDateAndTime)();
  BoolColumn get dirty => boolean().withDefault(const Constant(true))();
  BoolColumn get deleted => boolean().withDefault(const Constant(false))();

  @override
  Set<Column> get primaryKey => {rebuildSetId, partItemId, colorId};
}

/// A set's spare / extra parts (the "just in case" pieces LEGO ships with),
/// snapshotted from the catalog at add-time. Kept in its OWN table because the
/// catalog stores a spare as a separate row from the same build part (its PK
/// includes `is_spare`), so a spare and a build part can share a (part, colour)
/// key — merging them into [RebuildParts] would collide on the primary key.
/// [haveQty] is device-local "extras I found" tracking (a countable bonus,
/// excluded from build completion); this table is NOT cloud-synced — the needed
/// side is re-derived from the catalog on any device.
@DataClassName('RebuildExtraPartRow')
class RebuildExtraParts extends Table {
  TextColumn get rebuildSetId => text()();
  IntColumn get partItemId => integer()();
  IntColumn get colorId => integer()();
  IntColumn get neededQty => integer().withDefault(const Constant(0))(); // spare quantity
  IntColumn get haveQty => integer().withDefault(const Constant(0))();
  TextColumn get partName => text().withDefault(const Constant(''))();
  TextColumn get partNum => text().nullable()();
  IntColumn get partCatId => integer().nullable()();
  TextColumn get categoryName => text().nullable()();
  TextColumn get colorName => text().nullable()();
  TextColumn get colorRgb => text().nullable()();
  TextColumn get imageUrl => text().nullable()();

  @override
  Set<Column> get primaryKey => {rebuildSetId, partItemId, colorId};
}

@DataClassName('RebuildMinifigRow')
class RebuildMinifigs extends Table {
  TextColumn get rebuildSetId => text()();
  IntColumn get minifigItemId => integer()(); // catalog minifigs.item_id
  IntColumn get neededQty => integer().withDefault(const Constant(0))();
  IntColumn get haveQty => integer().withDefault(const Constant(0))();
  TextColumn get name => text().withDefault(const Constant(''))();
  TextColumn get imageUrl => text().nullable()();
  DateTimeColumn get updatedAt => dateTime().withDefault(currentDateAndTime)();
  BoolColumn get dirty => boolean().withDefault(const Constant(true))();
  BoolColumn get deleted => boolean().withDefault(const Constant(false))();

  @override
  Set<Column> get primaryKey => {rebuildSetId, minifigItemId};
}

@DataClassName('VerificationRow')
class Verifications extends Table {
  TextColumn get id => text()();
  TextColumn get rebuildSetId => text()();
  IntColumn get setItemId => integer()();
  RealColumn get completionPct => real().withDefault(const Constant(0))();
  IntColumn get partsNeeded => integer().nullable()();
  IntColumn get partsFound => integer().nullable()();
  IntColumn get minifigsNeeded => integer().nullable()();
  IntColumn get minifigsFound => integer().nullable()();
  TextColumn get flags => text().withDefault(const Constant('{}'))(); // json
  TextColumn get notes => text().nullable()();
  DateTimeColumn get verifiedAt => dateTime().withDefault(currentDateAndTime)();
  DateTimeColumn get updatedAt => dateTime().withDefault(currentDateAndTime)();
  BoolColumn get dirty => boolean().withDefault(const Constant(true))();
  BoolColumn get deleted => boolean().withDefault(const Constant(false))();

  @override
  Set<Column> get primaryKey => {id};
}

@DriftDatabase(
    tables: [RebuildSets, RebuildParts, RebuildMinifigs, Verifications, RebuildExtraParts])
class AppDatabase extends _$AppDatabase {
  AppDatabase() : super(driftDatabase(name: 'brickback'));

  /// For tests: back the DB with any executor (e.g. an in-memory NativeDatabase).
  AppDatabase.forTesting(super.executor);

  @override
  int get schemaVersion => 4;

  @override
  MigrationStrategy get migration => MigrationStrategy(
        onCreate: (m) => m.createAll(),
        // v1 -> v2: "group by type" + the extras (spares) list.
        // v2 -> v3: per-part counting step.
        // v3 -> v4: device-local F4 offline-image completeness marker.
        onUpgrade: (m, from, to) async {
          if (from < 2) {
            await m.addColumn(rebuildParts, rebuildParts.categoryName);
            await m.createTable(rebuildExtraParts);
          }
          if (from < 3) {
            await m.addColumn(rebuildParts, rebuildParts.stepQty);
          }
          if (from < 4) {
            await m.addColumn(rebuildSets, rebuildSets.imagesCachedAt);
          }
        },
      );
}
